src/main/java/com/palantir/gradle/gitversion/GitVersionCacheService.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.gradle.gitversion;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.internal.storage.file.FileRepository;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+
+public abstract class GitVersionCacheService implements BuildService<BuildServiceParameters.None> {
+
+    private final Timer timer;
+    private Map<String, VersionDetails> versionDetailsMap;
+
+    public GitVersionCacheService() {
+        timer = new Timer();
+        versionDetailsMap = new HashMap<>();
+    }
+
+    public final String getGitVersion(String project, Object args) {
+        File gitDir = getRootGitDir(new File(project));
+        GitVersionArgs gitVersionArgs = GitVersionArgs.fromGroovyClosure(args);
+        String key = gitDir.toString() + "|" + gitVersionArgs.getPrefix();
+        if (versionDetailsMap.containsKey(key)) {
+            return versionDetailsMap.get(key).getVersion();
+        }
+        Git git = gitRepo(gitDir);
```

Comment:
I wonder if this could just be pushed down into `VersionDetailsImpl`? For a couple of reasons:
1. It seems odd to initialize a JGit object in this class just to inject it into the VersionDetailsImpl instance
2. If we ever do decide to eliminate JGit, it would be good to have the usage limited in scope to make refactoring easier

On the _other hand_, I wonder if the `Git` class is thread-safe? The JGit javadoc doesn't really specify and a quick search indicates... maybe? If it is, then I wonder if it would make sense to cache a single instance here in this class instead of creating one instance per VersionDetailsImpl.

My preference would probably be to move this inside `VersionDetailsImpl` and instead pass `gitDir` as a constructor arg; a quick glance at `VersionDetailsImpl` seems to indicate that it is using _both_ JGit and Native Git to compute the description, which seems a little broken (see e.g.: https://github.com/palantir/gradle-git-version/blob/develop/src/main/java/com/palantir/gradle/gitversion/VersionDetailsImpl.java#L54-L60)

Reply:
Fixed by passing `gitDir` to the constructor rather than `Git` object. @bjlaub Do you want me to delete the Native git part as well?